docs/proposals/20220210-md-node-label-sync.md

```diff
@@ -0,0 +1,164 @@
+
+```
+node.cluster.k8s.io/*
+```
+
+Labels placed on the Machine that match the above prefix will be authoritative -- Cluster API will act as the ultimate authority for anything under the prefix. Labels with the above prefix should be set only within Cluster API CRDs.
+
+In addition to the above, CAPI will support one-time application of the following labels during Machine creation:
+
+```
+node-role.kubernetes.io/*
+node-restriction.kubernetes.io/*
+```
+
+The `node-role.kubernetes.io` prefix has been used widely in the past to identify the role of a node (e.g. `node-role.kubernetes.io/worker=''`). For example, `kubectl get node` looks for this specific label when displaying the role to the user.
+
+The use of the `node-restriction.kubernetes.io/*` prefix is recommended in the [Kubernetes docs](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#noderestriction) for things such as workload placement.
+
+It's important to note that the these specific labels are *not* kept in sync, but rather applied only once after the Machine and its corresponding Node are first created. Taking authoritative ownership of these prefixes would preclude other entities from adding and removing such labels on the Node. 
+
+### User Stories
+
+#### Story 1
+
+As a cluster admin/user, I would like a declarative and secure means by which to assign roles to my nodes. For example, when I do `kubectl get nodes`, I want the ROLES column to list my assigned roles for the nodes.
+
+#### Story 2
+
+As a cluster admin/user, for the purpose of workload placement, I would like a declarative and secure means by which I can add/remove labels on groups of nodes. For example, I want to run my outward facing nginx pods on a specific set of nodes (e.g. md-secure-zone) to reduce exposure to the company's publicly trusted server certificates.
+
+### Implementation Details/Notes/Constraints
+
+#### Label propogation from MachineDeployment to Machine
+
+Utilize CAPI metadata Propagation. CAPI supports [metadata propagation](https://cluster-api.sigs.k8s.io/developer/architecture/controllers/metadata-propagation.html?highlight=propagation#metadata-propagation) across core API resources. For the MachineDeployment types, template labels propagate to MachineSets top-level and MachineSets template metadata:
+```
+MachineDeployment.spec.template.metadata.labels => MachineSet.labels, MachineSet.spec.template.metadata.labels
+```
+Similarly, on MachineSet, template labels propagate to Machines:
+```
+MachineSet.spec.template.metadata.labels => Machine.labels
+```
+
+As of this writing, changing the `MachineDeployment.spec.template.metadata.labels` will trigger a rollout. This is undesirable. CAPI will be updated to instead ignore updates of labels that fall within the CAPI prefix. There is precedence for this with the handling of `MachineDeploymentUniqueLabel`. 
```

Comment:
Let's consider the following.
* A MD is initially created with some "node labels"
* Labels are propagated MD => MS => Machine
* Machine controller will propagate the "node labels" from the Machine to the node in the workload cluster

Now we modify one of those node labels on the MachineDeployment. With our changes to the MD controller this won't trigger a rollout of MachineSets / Machines (which is intended) but it also won't propagate the changed node labels to the workload nodes as the MachineSet and the Machine is not rolled out and thus the machine controller doesn't know that it has to sync the changed node labels.

I assume this is not intended. Do we need some kind of "inline patching" in the MD and MS controllers? I.e. the MD controller patches the existing MachineSets and the MachineSet controller patches the Machines if node label change? (instead of rolling them out like for all other changes)

Reply:
Yeh I think this need to be rephrased a bit to clarify we'll introduce in-place reconciliation from the MD->MS-Machines for these labels. Implementation wise this need to be solved generically adding first class support for [in-place propagation which also affect other scenarios](https://github.com/kubernetes-sigs/cluster-api/issues/5880).

Reply:
One thing that we have to keep in mind is that there might be user expectations today that changing labels/annotations triggers rollouts.

I think it's sometimes used as a workaround because rolloutAfter isn't available yet on MDs. We even mention it in our documentation: https://cluster-api.sigs.k8s.io/tasks/upgrading-clusters.html?highlight=rolloutAfter#how-to-schedule-a-machine-rollout

Resolution could be to get the MachineDeployment rolloutAfter in place before merging inline patching of labels/annotations.

Reply:
Is there a ticket for  MachineDeployment rolloutAfter? It will be a blocker for the inplace propagation of labels and in general seems like a good feature to have.

Reply:
I'm not sure I follow how is rolloutAfter related? I'd expect us to do this in a way we preserve existing behaviour.

>I assume we would do it for all labels not only the ones for nodes.

Ah I see because of this assumption. I'm not 100% sure on this assumption, may be this is another arg to have a specific field in the API for Node labels.

Reply:
Yup, sounds like a good reason to have a separate field, I think otherwise there is no way to preserve existing behavior.